File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Ports/EventBus.swift
// ROLE: port -- a publish/subscribe channel for what the connection edge is doing.
// IMPLEMENTED BY: SimpleEventBus; FakeEventBus.
// USED BY: BridgeServer, which emits after every state transition, and the launcher, which subscribes.
// Handlers run on the publisher's thread, the accept loop's; a subscriber that touches AppKit must marshal to the main thread.

public enum ServerState: String, Equatable, Sendable {
	case stopped
	case listening
	case sessionActive
}

/// `endpoint` is nil when stopped.
public struct ServerStatus: Equatable, Sendable {
	public let state: ServerState
	public let endpoint: String?

	public init(state: ServerState, endpoint: String?) {
		self.state = state
		self.endpoint = endpoint
	}
}

public enum BridgeEvent: Equatable, Sendable {
	case serverStatus(ServerStatus)
}

public typealias SubscriptionToken = String

public protocol EventBus: AnyObject {
	func subscribe(_ handler: @escaping (BridgeEvent) -> Void) -> SubscriptionToken

	/// Safe to call twice, and with a token already removed.
	func unsubscribe(_ token: SubscriptionToken)

	func emit(_ event: BridgeEvent)
}

File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/ports/event_bus.py
# nvdaMcpBridge domain ports -- the EventBus port.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: domain port, a pub/sub channel shared by the domain and adapters.
# IMPLEMENTED BY: adapters/simple_event_bus.py.
# USED BY: BridgeServer as emitter and BridgeDialog as subscriber.

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Callable

from ..entities.bridge_events import BridgeEvent, BridgeEventType

EventHandler = Callable[[BridgeEvent], None]

SubscriptionToken = str


class EventBus(ABC):
	"""Handlers run synchronously on the publishing thread; a UI handler must marshal itself."""

	@abstractmethod
	def subscribe(self, event_type: BridgeEventType, handler: EventHandler) -> SubscriptionToken:
		"""Returns a token for unsubscribe()."""

	@abstractmethod
	def unsubscribe(self, token: SubscriptionToken) -> None:
		"""Safe to call when already removed."""

	@abstractmethod
	def emit(self, event: BridgeEvent) -> None:
		"""Deliver *event* to every handler registered for its type."""
